crates/runtime/src/instance.rs

```diff
@@ -843,4 +854,71 @@ impl InstanceHandle {
             instance: self.instance,
         }
     }
+
+    /// Creates an internal snapshot of this instance's memory and tables
+    /// to be restored at a later time.
+    ///
+    /// Should only be called once.
+    #[cfg(target_os = "linux")]
+    pub fn create_snapshot(&mut self) -> Result<(), anyhow::Error> {
+        let instance = self.instance_mut();
+        if !instance.is_reusable {
+            anyhow::bail!("instance is not reusable");
+        }
+
+        for memory in instance.memories.values_mut() {
+            memory.create_snapshot()?;
+        }
+
+        for table in instance.tables.values_mut() {
+            table.create_snapshot()?;
+        }
+
+        for index in instance.module.globals.keys() {
+            let value = unsafe { std::ptr::read(instance.defined_or_imported_global_ptr(index)) };
+            instance.saved_globals.push((index, value));
+        }
+
+        Ok(())
+    }
+
+    /// Resets this instance's memory, tables and globals to their initial
+    /// state from when the [`InstanceHandle::create_snapshot`] was called.
+    ///
+    /// Can be called any number of times.
+    #[cfg(target_os = "linux")]
+    pub fn restore_snapshot(&mut self) -> Result<(), anyhow::Error> {
+        let instance = self.instance_mut();
+        if !instance.is_reusable {
+            anyhow::bail!("instance is not reusable");
+        }
+
+        for index in (0..instance.memories.len() as u32).map(DefinedMemoryIndex::from_u32) {
+            let memory = &mut instance.memories[index];
+            memory.restore_snapshot()?;
+
+            let vmmemory = memory.vmmemory();
+            instance.set_memory(index, vmmemory);
```

Comment:
I'm pretty sure this is necessary for correctness, but when I comment it out nothing fails. Any idea how I could write a test which would verify that this is here?